lib/webmock/util/hash_counter.rb

```diff
@@ -5,22 +5,36 @@
 module WebMock
   module Util
     class HashCounter
-      attr_accessor :hash
+      attr_accessor :hash, :array
```

Comment:
@johngallagher I believe we should separate the array of requests from the HashCounter. The HashCounter has a specific, single responsibility, and adding an array of requests expands its scope beyond its intended purpose.

Instead, I suggest moving the ordered list of requests up to the RequestRegistry object. This way, the RequestRegistry would manage both the HashCounter and the array of requests, maintaining a clearer separation of concerns.

What are your thoughts on this restructuring? It would allow the HashCounter to remain focused on its core functionality.

Reply:
@bblimke yup, that would work better. I do think that having all this cruft in the `RequestRegistry` might clutter it up, so it might be a `RequestStore` or something similar... although that's literally what `RequestRegistry` does I guess... so I'll look into that. Thanks! Great idea.

Reply:
@bblimke aha - I remember why I didn't put it into a separate object now. 💡 

It's because of the design - `HashCounter` is exposed directly in lots of places like so:

https://github.com/johngallagher/webmock/blob/088aa2678f7efb34e910e8b430b54980283b9485/lib/webmock/http_lib_adapters/curb_adapter.rb/#L61

Personally, I'd rather encapsulate how the registry stores requests and hide that from clients.

That would mean we could have one store, two or more. But given how it's done, without a huge refactoring, I don't see how we'd get this in.

I'd be happy to take on that refactoring, to be clear - I love me some refactoring!

It'd be:

```diff
- WebMock::RequestRegistry.instance.requested_signatures.put(request_signature)
+ WebMock::RequestRegistry.instance.put(request_signature)
```

And the `requested_signatures` would be entirely private. We could have two different store classes that way and do something like (sketching it out roughly):

```ruby
class WebMock::RequestRegistry
  def initialize
    @hash_counter_store = HashCounterStore.new,
    @array_store = ArrayStore.new
  end

  def put(request_signature)
    @hash_counter_store.put(request_signature)
    @array_store.put(request_signature)
  end

  def get(request_signature)
    @array_store.get(request_signature)
  end

  def to_a
    @array_store.to_a
  end

  def to_h
    @hash_counter_store.to_h
  end
  # ...
end
```

This would give us duck typing for stores, make their internals private and allow us to reach into the relevant store when we wanted to pluck out requests that were stored in a specific format.

WDYT?

Reply:
@bblimke **nudge** :)

Reply:
@johngallagher Thanks for the reminder. I agree that RequestRegistry internals should be private, with no other object having direct access to the hash counter or the array. I like your solution with two separate stores: hash_counter_store and array_store.

I'm not sure if this is a breaking change, since RequestRegistry isn't supposed to be used directly. It's not part of WebMock::API, which is the only versioned interface. However, there's a likelihood that some people have accessed RequestRegistry directly.

I don't mind releasing a new major version though, if we decide it's necessary.